### PR TITLE
syncer: add retry for binlog event streamer

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
Also allow error retry for remote binlog reader

### What is changed and how it works?
- For remote binlog puller, allow auto-retry pullEvent error but restrict to only once per specified interval (currently 1min). 

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

Related changes

### CODE REVIEW

**Review by lance6716 — APPROVED** [review]

File: syncer/streamer_controller.go [hunk 1]
```diff
@@ -476,3 +497,33 @@ func isConnectionRefusedError(err error) bool {
...
+	CanRetry(error) bool
+}
+
+type alwaysRetryStrategy struct{}
+
+func (s alwaysRetryStrategy) CanRetry(error) bool {
+	return true
+}
+
+// maxIntervalRetryStrategy allows retry the duration since last retry is bigger than specified interval.
```
Comment (Ehco1996) [inline syncer/streamer_controller.go:512]: maybe this comment is better?

`allow retry when the retry interval is greater than the set interval.`

File: syncer/streamer_controller.go [hunk 2]
```diff
@@ -120,10 +124,30 @@ type StreamerController struct {
...
+	// let local binlog also return error to avoid infinity loop
+	failpoint.Inject("GetEventError", func() {
+		strategy = &maxIntervalRetryStrategy{
+			interval: minErrorRetryInterval,
+		}
+	})
 	streamerController := &StreamerController{
 		initBinlogType:    binlogType,
 		currentBinlogType: binlogType,
+		retryStrategy:     strategy,
```
Comment (Ehco1996) [inline syncer/streamer_controller.go:150]: can we use func `NewStreamerController` to init streamerController here ?

File: tests/incremental_mode/run.sh [hunk 3]
```diff
@@ -168,7 +169,8 @@ function run() {
 	check_port_offline $WORKER1_PORT 20
 	check_port_offline $WORKER2_PORT 20
 
-	export GO_FAILPOINTS="github.com/pingcap/dm/syncer/WaitUserCancel=return(8)"
+	# only mock pull binlog failed once
+	export GO_FAILPOINTS="github.com/pingcap/dm/syncer/WaitUserCancel=return(8);github.com/pingcap/dm/syncer/GetEventError=1*return"
```
Comment (Ehco1996) [inline tests/incremental_mode/run.sh:173]: Can you please tell me what the usage of `1 * retrun` ?

**Review by Ehco1996 — COMMENTED** [review]
restLGTM

File: syncer/streamer_controller.go (hunk 2 above)
Comment (glorv) [reply syncer/streamer_controller.go:150]: I don't get the point, this function is `NewStreamerController` 😕️

File: tests/incremental_mode/run.sh (hunk 3 above)
Comment (glorv) [reply tests/incremental_mode/run.sh:173]: Only trigger this error once. See: https://github.com/pingcap/failpoint#key-concepts

**Review by Ehco1996 — APPROVED** [review]

**Pull request merged**